test-framework/test_framework.go

```diff
@@ -0,0 +1,226 @@
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/errors"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/parser"
+	"github.com/onflow/cadence/runtime/sema"
+	"github.com/onflow/cadence/runtime/stdlib"
+	"github.com/onflow/cadence/runtime/tests/utils"
+)
+
+// This Provides utility methods to easily run test-scripts.
+// Example use-case:
+//   - To run all tests in a script:
+//         RunTests("source code")
+//   - To run a single test method in a script:
+//         RunTest("source code", "testMethodName")
+//
+// It is assumed that all test methods start with the 'test' prefix.
+
+const testFunctionPrefix = "test"
+
+type Results map[string]error
+
+func RunTest(script string, funcName string) error {
+	_, inter := parseCheckAndInterpret(script)
+	_, err := inter.Invoke(funcName)
+	return err
+}
+
+func RunTests(script string) Results {
+	program, inter := parseCheckAndInterpret(script)
+
+	results := make(Results)
+
+	for _, funcDecl := range program.FunctionDeclarations() {
+		funcName := funcDecl.Identifier.Identifier
+
+		if strings.HasPrefix(funcName, testFunctionPrefix) {
+			_, err := inter.Invoke(funcName)
+			results[funcName] = err
+		}
+	}
+
+	return results
+}
+
+func parseCheckAndInterpret(script string) (*ast.Program, *interpreter.Interpreter) {
+	program, err := parser.ParseProgram(script, nil)
+
+	checker, err := newChecker(program)
+	if err != nil {
+		panic(err)
+	}
+
+	err = checker.Check()
+	if err != nil {
+		panic(err)
+	}
+
+	// TODO: validate test function signature
+	//   e.g: no return values, no arguments, etc.
+
+	inter, err := newInterpreterFromChecker(checker)
+	if err != nil {
+		panic(err)
+	}
+
+	err = inter.Interpret()
+	if err != nil {
+		panic(err)
+	}
+
+	return program, inter
+}
+
+func newInterpreterFromChecker(checker *sema.Checker) (*interpreter.Interpreter, error) {
+	predeclaredInterpreterValues := stdlib.BuiltinFunctions.ToInterpreterValueDeclarations()
+	predeclaredInterpreterValues = append(predeclaredInterpreterValues, stdlib.BuiltinValues.ToInterpreterValueDeclarations()...)
+	predeclaredInterpreterValues = append(predeclaredInterpreterValues, stdlib.HelperFunctions.ToInterpreterValueDeclarations()...)
+
+	return interpreter.NewInterpreter(
+		interpreter.ProgramFromChecker(checker),
+		checker.Location,
+		interpreter.WithStorage(interpreter.NewInMemoryStorage(nil)),
+		interpreter.WithTestFramework(NewEmulatorBackend()),
+		interpreter.WithPredeclaredValues(predeclaredInterpreterValues),
+		interpreter.WithImportLocationHandler(func(inter *interpreter.Interpreter, location common.Location) interpreter.Import {
+			switch location {
+			case stdlib.CryptoChecker.Location:
+				program := interpreter.ProgramFromChecker(stdlib.CryptoChecker)
+				subInterpreter, err := inter.NewSubInterpreter(program, location)
+				if err != nil {
+					panic(err)
```

Comment:
Should this be wrapped in `errors.NewUnexpectedError`?

Reply:
We could, but doesn't really need to. The wrapping was introduced specially for the cadence <-> fvm interface, so the fvm can 'fail transaction' vs 'stop the network' depending on the error type. In tests, we don't really have that requirement. All errors are recovered at the entry point and the test is failed.